snapcraft/internal/project_loader/_templates.py

```diff
@@ -0,0 +1,218 @@
+# -*- Mode:Python; indent-tabs-mode:nil; tab-width:4 -*-
+#
+# Copyright (C) 2018 Canonical Ltd
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License version 3 as
+# published by the Free Software Foundation.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+import contextlib
+import copy
+import functools
+import jsonschema
+import logging
+import os
+from typing import Any, Dict, List, Set  # noqa: F401
+import yaml
+
+from snapcraft import formatting_utils
+from snapcraft.internal import common
+from . import errors
+
+logger = logging.getLogger(__name__)
+
+# Templates are evaluated before the snapcraft.yaml is validated, so we need to handle
+# our own subset of validation ourselves (covering only the usage of templates).
+template_schema = {
+    "type": "array",
+    "minitems": 1,
+    "uniqueItems": True,
+    "items": {"type": "string"},
+}
+
+
+def apply_templates(yaml_data: Dict[str, Any]) -> Dict[str, Any]:
+    """Apply all templates.
+
+    :param dict yaml_data: Loaded, unprocessed snapcraft.yaml
+    :returns: Modified snapcraft.yaml data with templates applied
+    """
+    # Don't modify the dict passed in
+    yaml_data = copy.deepcopy(yaml_data)
+
+    # Get the base being used for this project (defaults to "core")
```

Comment:
core is not a base :-)
I would default to nothing and if needed, introduce this later

Reply:
IOW, let's tie this to the use of bases.

Reply:
Remember that today, you can run snapcraft on arch and if no stage packages or build-packages are introduced, chances are, it might just work. This would introduce another vector of confusion.

Reply:
I disagree on this point, for a few reasons:

1. Remote parts are a problem today, with a ton of assumed knowledge and documentation issues. They are hard to use, and templates will be better. They are also very widely used (desktop helpers in particular).
2. Templates are the only option for folks using remote parts that move to bases (as remote parts are not supported). By not allowing folks to transition to templates before they transition to using bases, we make the move to using bases quite a lot more work than is actually necessary.
3. Running snapcraft today on things other than Ubuntu is not something we support without bases. Making decisions based on how snapcraft runs on other distros today sounds a bit like support :wink: .

Unless supporting templates before bases actually makes our lives harder (I don't think it does), I think it's a good idea. We can chat about this more in-person, though.

Reply:
We chatted about this in person, and the big concern is simply that the core16 base, once it exists, cannot guarantee that it will continue building on trusty as well, which supporting templates without bases would entail (i.e. defaulting to core16 templates). As a result, I've started raising an exception if not base is specified.